### ISSUE TITLE
false negative when matching optional unnamed parameters

### ISSUE THREAD

**Issue by vlukashov** [pillarjs/path-to-regexp#206]
The regexp for the `/user(s)?/:user` path does match the `/user/sam` path with `path-to-regexp` v2.4.0, but not with the newer versions. Could you please confirm whether or not this is an expected behavior? I do not see relevant info in the README file.

With `path-to-regexp` v2.4.0:
```js
const regexp = pathToRegexp('/user(s)?/:user');
// keys = [{ name: 0, ... }, { name: 'user', ... }]

regexp.exec('/user/sam');
//=> Array (3) [ '/user/sam', undefined, 'sam', index: 0, input: '/user/sam' ]
```

With `path-to-regexp` v3.0.0 and newer:
```js
const regexp = pathToRegexp('/user(s)?/:user');
// keys = [{ name: 0, ... }, { name: 'user', ... }]

regexp.exec('/user/sam');
//=> null
```
I would expect the `/user/sam` path to match the regexp since the unnamed `(s)` group is optional.

I've added more resent versions of `path-to-regexp` into the [Express Route Tester](https://vlukashov.github.io/express-route-tester/) so that this can be easily verified.

**Comment by blakeembrey** [comment]
It's caused by https://github.com/pillarjs/path-to-regexp/blob/master/History.md#300--2019-01-13, you'll need to escape the `\\r` so it's not treated as a prefix (or use `whitelist: '/.'`). It's not an ideal system so I can investigate how to make it better, but the escape mechanism will allow you partial matching.

**Comment by blakeembrey** [comment]
@paulmelnikow I see you've also 👍'd this, would love to align on a spec to define the behavior. Personally, this feels like it's largely caused by a lack of opening and closing tokens for path segments which is an issue with the original `path-to-regexp` syntax. By trying to keep with that syntax, I've introduced confusion. This is because, for the whole thing to be reversal and support repeated/optional parameters, you need to understand the prefix token. 

For example, `/foo/:bar?` expects `/foo` to match (`/:bar?` is considered the entire token). This is more true once you have repeated parameters also. The only way I can think of, to get around the obscurity of this feature, is to have `:bar` exclude the prefix token and adopt a clearer syntax for denoting prefix characters. E.g. `/foo{/bar}` (not sure if the lack of `:` makes it unclear now). 

Alternatively, we keep `whitelist: './'` as the default and people wanting things like `/:attr1-:attr2?` won't work correctly. Those get awkward though with `compile`, since you don't particularly expect to be required to put the `-` prefix onto each token you wish to popular. With the alternative syntax above, it'd be possible via `/:attr1{-attr2}?` (or maybe `{-:attr2}?`).

Happy to open a new issue with the spec in plain English if anyone wants to chime in!

**Comment by vlukashov** [comment]
Thanks for the explanation! It is clear to me now that the current behavior is expected, though not ideal. 

It would have helped me if I saw a bit more thorough explanation in the change log. Something along the lines of

*Now `path-to-regexp` supports routes like `/:user/:repo/:branch*`, but that required a breaking change. Starting from the version `3.0.0` routes like `/user(s)?/:user` would not work without modifications: if the character before the unnamed optional group is not a `/` then it needs to be escaped: `/use\\r(s)?/:user`. More details in https://github.com/pillarjs/path-to-regexp/pull/176.*

For my part this issue can be closed.

**Comment by frenzzy** [comment]
In my opinion `/(user|users)/:user` looks better than `/use\\r(s)?/:user`, just my 2 cents :)
+1 for more info about braking changes in the changelog

**Comment by blakeembrey** [comment]
Opened an exploratory PR in https://github.com/pillarjs/path-to-regexp/pull/207. With this change, I can also re-introduce `whitelist: './'` by default for prefix characters. This would minimize the breaking change overall and it'd purely be an addition for the people who wanted non-whitelist prefixes. Please feel free to move the discussion to that PR (I haven't done the whitelist change right now). It would also make the routes act mostly like 0.x - <3.x again.

**Comment by blakeembrey** [comment]
Fixed with v6.
